Less strict size check in scatter

redmvc opened the pull request:
Fixes #63265.

This relaxes the constraint in scatterlike functions where all dimensions of the index tensor must be smaller than or equal to the size of the same dimensions of the source tensor. This is done by creating a separate pathway specific for this use case and then rebuilding the implied position of a given entry in the index tensor then constructing the corresponding position in the source tensor and recalculating the offset for it. Both CUDA and CPU are implemented and should work identically.

I added tests to check that everything is still working, but there were a couple of failures when the tests were run on `device = 'cuda'` and `dtype` was a float of some sort. In those cases, there seems to be a small (usually sub 1%) fraction of the tested entries that have a small (always sub 1.5%) relative difference between actual and expected values. I ran the tests locally without using the "src smaller than index" flag and still ran into those differences, though, so I'm not sure they were introduced by my changes.

I'm a first-time contributor to pyTorch so this probably has substantial issues or bits where I did not follow best practices. Nevertheless, it seems to work, and at least for generic use cases I tested does not seem much slower, though it is _somewhat_ slower as it requires some extra computation (O(ndim)). For normal use cases (where index has no dimension larger than the same dimension in source) the original code is used instead, so the effect of this change on existing applications should be minimal.

redmvc committed:
Remove constraint from checks

Removes the index.size(d) > src.size(d) constraint from the "checks" file.

redmvc committed:
Update documentation

Updates the documentation to reflect the removal of the constraint.

redmvc committed:
Case d == dim

Permits index.size(dim) to be greater than src.size(dim).

redmvc committed:
Case d != dim, dim == self.dim() - 1

Permits index.size(d) to be greater than src.size(d) for any d, provided dim is the last dimension of the output tensor.

redmvc committed:
Revert "Case d != dim, dim == self.dim() - 1"

This reverts commit b6dbb66e486691f98476f1752fcdf25c53240688.

redmvc committed:
Actually remove constraint from checks

Removes the actual code of the checks that constrains index.size(d) <= src.size(d).

redmvc committed:
Case CUDA

Implemented the relevant logic for the CUDA version of the functions.

redmvc committed:
Revert "Case CUDA"

This reverts commit 45b212dceb9cb426c82668c54b0c10d0d2a2e985.

redmvc committed:
Case CUDA

Implemented the relevant logic for the CUDA version of the functions.

redmvc committed:
Fix source of tensor sizes

Fixes the erroneous assumptions about how the iterator stores sizes and turned the variables into input to the function.

redmvc committed:
Fix variable types

Fixes the variable types of the pointer offsets.

redmvc committed:
Rename vector variables

Renames the variables that are taken in their vector forms to clarify this.

redmvc committed:
Create index_sizes variable

Create a variable for index.sizes() as it is referred to twice in the functions.

redmvc committed:
Create variable to track this use case

Creates a variable that tracks whether the index tensor has any dimension bigger than the source dimension so as to not apply this whole logic unless necessary.

redmvc committed:
Create variable to track source strides

Creates a variable that tracks source strides for later use.

redmvc committed:
Add src_strides as input to kernel function

Replace the assignment of source tensor strides in the kernel function from the iterator to use it as input instead.

redmvc committed:
Edit source restriding

Edits the restriding of the source tensor when the index has any dimensions larger than it.

redmvc committed:
Replace reference to src.sizes()

Replaces the references to src.sizes() in the kernel function call with src_sizes.

redmvc committed:
Move ndim variable declaration to within relevant scope

redmvc committed:
Fix function call

redmvc committed:
Fix typo

redmvc committed:
Add temporary debugging check

redmvc committed:
Revert "Add temporary debugging check"

This reverts commit a1b6d3ea77a28174df3e5bd2bf9ca1c727946928.

redmvc committed:
Add temporary debugging check

redmvc committed:
Revert "Edit source restriding"

This reverts commit 0ef9449391a8bbdca7a4ec8d514276fc7e539c51.

redmvc committed:
Change source tensor restride

Changes the restriding of the source tensor to reduce all strides to 0 in this use case.

redmvc committed:
Revert "Move ndim variable declaration to within relevant scope"

This reverts commit 2713f8a46f8b4658a8a1c2c65e55085b7dbba207.

redmvc committed:
Add debugging variable

redmvc committed:
Add debugging message

redmvc committed:
Update debugging messages

redmvc committed:
Remove debugging messages

redmvc committed:
Remove unnecessary namespace call

redmvc committed:
Pass index_strides as argument to the function

redmvc committed:
Turn input arrays into references

redmvc committed:
Remove debugging variable

redmvc committed:
Adjust offsets to account for word size

redmvc committed:
Remove references

redmvc committed:
Separate this case into its own function

Separates the kernel function into two, one for this use case and one for the default use case, for simplicity and clarity.

redmvc committed:
Fix comment

redmvc committed:
Remove unused variable

redmvc committed:
Rename sizes variables

Renames the _sizes variables within the large index kernel function to the less ambiguous _shape.

redmvc committed:
Add explicit cuda memory allocation for variables

redmvc committed:
Typo fix

redmvc committed:
Remove unused variables

redmvc committed:
Rename case detection variable

Renames the index_larger_than_src variable to index_larger_than_src_in_scatter to clarify usage, as this variable is not used at all for gather.

redmvc committed:
Fix offset-calculating variable types

redmvc committed:
Change src_offset variable type

Changes the variable type of src_offset to be the same as that of the other offsets.

redmvc committed:
Use assignment operator for src_offset

redmvc committed:
Remove redundant condition

redmvc committed:
Fix AT_DISPATCH macro

redmvc committed:
Explicitly set dimension variables to int

redmvc committed:
Fetch number of dimensions from shape array

redmvc committed:
Typo fix

redmvc committed:
Revert "Typo fix"

This reverts commit e9b18bac11000c66ea06dca97bce9bf772f0ce95.

redmvc committed:
Typo fix

redmvc committed:
Copy data directly from original pointer

redmvc committed:
Add explanatory comments to cuda version

redmvc committed:
Move ndim declaration

redmvc committed:
Remove functionality from the "TensorAssign" version

redmvc committed:
Typo fix

redmvc committed:
Restore condition to documentation of TensorAssign version

redmvc committed:
Get word size dynamically

redmvc committed:
Update documentation to reflect this use case

redmvc committed:
Revert changes to CPU version

redmvc committed:
Revert removal of functionality from TensorAssign

redmvc committed:
Add examples to the documentation

redmvc committed:
Case CPU

redmvc committed:
Fix function argument variable type

redmvc committed:
Typo fix

redmvc committed:
Typo fix

redmvc committed:
Fix calculation location

redmvc committed:
Fix address call in dim kernel function

redmvc committed:
Remove word size comments

redmvc committed:
Add explanatory comments to CPU file

redmvc committed:
Fix whitespace

redmvc committed:
Typo fix

redmvc committed:
Add tests

Edits the _test_scatter_base function to receive a new parameter, src_smaller_than_index, to test these cases, and adds versions of all existing tests that set this parameter to True to test functioning of the feature.

redmvc committed:
Lint

redmagnos commented:
So @mikaylagawarecki I'm not sure how the process works and whether you're waiting for all checks to pass before taking a first look (I'm new at this so don't know what the general procedures are ^^') but as far as I can tell there is one specific check that's failing - the tests I added for the feature. However, the failures are very small (as I mentioned in the original PR message), and at least when I ran the tests locally on a branch that doesn't have my changes, the existing tests still failed at similar rates.

Which does kind of leave me confused, though, as to how other branches/PRs pass all tests? Do they not always check the scatter/gather ops and only do this when those are changed, or something? Or is there something I should do on my end?

mikaylagawarecki commented on test/test_scatter_gather_ops.py:
```diff
@@ -98,7 +98,11 @@ def _test_scatter_base(self, fn, *, device, dtype, is_scalar, reduction,
+            sign = [1, 1, 1]
+            if src_smaller_than_index:
+                while(min(sign) > 0):
+                    sign = [random.choice([-1, 1]) if s > 1 else 1 for s in idx_size]
```
does this mean we don't generate cases where src is smaller than index on dim 0?

redmagnos replied:
No, I don't think so, unless I misunderstand your question. The `s` there is the size in a given dimension, so if `s = index.size(d) == 1` for some dimension `d` then it won't generate a `src` tensor that has size less than that along that dimension (since afaik there's no such thing as size `0` in a 3-dimensional tensor and this would break stuff). `s == 1` can only happen on `dim`, though, since all other dimensions in that generator are `>= 10`, so that loop is guaranteed to make at least one of the `src` dimension sizes smaller than the corresponding `index` dimension size.

mikaylagawarecki replied:
My bad, gotcha thanks!

mikaylagawarecki commented:
@redmagnos Thank you for contributing as well as your patience on the review! Regarding the tolerance issues on CUDA, that is likely because scatter/scatter_add/scatter_reduce use atomics on CUDA which are nondeterministic. It is interesting that you are seeing failures on these tests before the PR though as these tests should be running in CI. Let me try to reproduce this locally.

nikitaved commented:
Thank you, @redmagnos , for this work! I will definitely have a look at it. Scatter is nondeterministic, but you can make it deterministic by using a source with all-equal elements!

mikaylagawarecki commented on test/test_scatter_gather_ops.py:
```diff
@@ -85,7 +85,7 @@ def test_gather_backward_with_empty_index_tensor(self, device, dtype, sparse_gra
 
     def _test_scatter_base(self, fn, *, device, dtype, is_scalar, reduction,
-                           unique_indices=True, include_self=True):
+                           unique_indices=True, include_self=True, src_smaller_than_index=False):
```
One more thing, could you please add some sample inputs to the sample_inputs functions for the opinfos of scatter/scatter_add/scatter_reduce ([e.g. here](https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/common_methods_invocations.py#L8328)) where src is smaller than index for the case where `not requires_grad` (since we aren't supporting gradients for this case)? 

This will run more comprehensive tests (e.g. `python test/test_ops.py -k scatter` includes a test that your proposed changes work on [noncontiguous tensors](https://github.com/pytorch/pytorch/blob/master/test/test_ops.py#L575))

redmagnos replied:
Roger that!

Also, I am not sure these changes actually do work on noncontiguous tensors, I haven't tested them. I _think_ they should, if the logic works as I think it does, but I'll check.

redmagnos replied:
Testing manually with some small tensors the logic does seem to work for noncontiguous tensors, and I've added sample inputs that I _think_ do the thing, if I understood them correctly.

redmagnos replied:
Okay, I had _not_ understood it correctly actually and my logic won't work for it. I'll think about how to fix it.

redmagnos replied:
@mikaylagawarecki implemented the logic for CUDA, but as @nikitaved said we might want to go with the "intersection" interpretation of the request instead of the "wraparound" one, so I'll hold off on doing it for CPU until that's decided!

redmvc committed:
Fix comments in documentation

redmagnos commented on torch/_tensor_docs.py:
```diff
@@ -3433,6 +3448,37 @@ def callable(a, b) -> number
+
+            [[ 7,  8,  9],
+             [10, 11, 12]]])
+    >>> gen = torch.Generator()
+    >>> gen.manual_seed(123456)
+    >>> index = torch.empty((2, 3, 4), dtype = torch.long).random_(0, 3, generator = gen)
```
Fair enough! I used the generator there because I wanted the index to be reasonably big (24 elements in that case) and I thought it'd look clunkier to write all 24 elements out by hand.

redmagnos commented on aten/src/ATen/native/cuda/ScatterGatherKernel.cu:
```diff
@@ -154,7 +154,85 @@ struct _cuda_scatter_gather_internal_kernel {
+    // We send the relevant index and source metadata
+    // to the device
+    int ndim = index_shape.size();
+    int64_t *index_shape_device, *index_strides_device, *src_shape_device, *src_strides_device;
```
That's exactly the sort of thing I expected I'd miss/not know about, thank you! I'll look into changing that.

redmagnos replied:
Okay, so, looking into it, I am not quite sure which of the various caching allocators I should use. There seem to be several options, the most likely-looking one being the `CUDACachingAllocator` one, but looking through examples in the codebase to check for usage does not make it immediately obvious how to then copy the data from host to device. I don't suppose you'd know of some resource or guide or particularly enlightening snippet that could point at what the best practices are?

nikitaved replied:
Why not allocating a single buffer with a Tensor? You can get it's head with `t.data_ptr<type>()`.

redmagnos replied:
I'm not sure what you mean by that. What does seem necessary to me is a way to have an ordered list of strides and sizes for index and source, given that there's (as far as I could tell) no way to recover the information of what index a given offset is equivalent to in the original tensor with just the information otherwise provided to the loop. But there might be a way to access those values from the iterator itself or something in there that I missed, in which case passing it to device memory the way I did would be entirely unnecessary.

redmagnos commented on torch/_tensor_docs.py:
```diff
@@ -3433,6 +3448,37 @@ def callable(a, b) -> number
     tensor([[2.0000, 2.0000, 3.2300, 2.0000],
             [2.0000, 2.0000, 2.0000, 3.2300]])
 
+    >>> # Testing the case where index.size(d) > src.size(d)
```
Roger that

redmagnos commented on aten/src/ATen/native/cpu/ScatterGatherKernel.cpp:
```diff
@@ -128,6 +128,54 @@ struct _cpu_scatter_gather_dim_loop {
+        " with size ", index_upper_bound
+      );
+
+      // calculate the equivalent indices in the source tensor based on
```
As I understand it, from that comment, if we restrided the tensors in that way the entries of `index` that go beyond those in `src` would be ignored rather than looping back, which I think was the desired behaviour? Unless I'm missing something.

nikitaved replied:
Right, only indices in the "intersection" are considered.

mikaylagawarecki replied:
Hey @redmagnos, this is a good point! While the intersection is one possible interpretation that would be simpler to implement (which I believe is why the issue was labelled as a good first issue), you are right that the (much more technically complex) wraparound behavior that you have implemented seems to be what #63265 was suggesting judging by the code block under alternatives. 

Since the wraparound behavior doesn't seem to be working for noncontiguous tensors at the moment (per your comment below) perhaps we could hold off on fixing this for a bit while we determine which interpretation of the enhancement would best serve our users.

If you had a use-case for this wraparound behavior that motivated you to take up this issue, it would be great if you could share as well :)

redmagnos replied:
I don't, actually; I assumed the original poster themself would have a reason, but the reason I picked this was more "it was marked as a good first issue and no one else seemed to be actively working on it".

With respect to noncontiguous tensors, I implemented something that seems to work for them and had been testing them earlier today, so since the work is already done I'll push it and you guys can take a look. And in case the "intersection" solution is the preferred one after all, it should be reasonably easy to pivot in that direction (since it is in fact just a matter of restriding with no change in the logic elsewhere at all).

redmvc committed:
Set documentation example index tensors' values manually

redmvc committed:
Add sample inputs

Adds samples with src smaller than index to the sample_inputs functions for scatter, scatter_add, and scatter_reduce.

redmvc committed:
Lint

redmagnos commented:
So I've been scratching my head about the Mysterious Failures in various tests here and finally saw the message the bot created a while ago talking about how I should be rebasing onto `viable/strict` rather than `master`, which might explain at least some of it? If so, whoops. I couldn't find this mentioned in the contributing docs but I might've missed it, and I definitely missed it the first time the bot commented here.

If that's the main reason for (at least some of) these failures then probably when I rebase onto `viable/strict` most of those failures should disappear. Hopefully!

janeyx99 commented:
@redmagnos I suspect you may have rebased onto your fork's viable/strict or an outdated viable/strict, which is probably behind the current viable/strict in pytorch/pytorch. To rebase onto our viable/strict next time, I recommend doing something like:
```
git fetch --all    # to pull the most recent viable/strict
git rebase upstream/viable/strict OR git rebase origin/viable/strict     # you can check your .git/config to see whether upstream or origin points to pytorch/pytorch
```

In this case, I would recommend using `git cherry-pick` to select the commits you are trying to push onto a fresh branch. If you want to base that onto pytorch/pytorch's viable/strict, you can do something similar to:
```
git fetch --all
git checkout upstream/viable/strict 
git cherry-pick <commit-a> <commit-b> ...
git checkout -b less-strict-scatter-size-check
...
```
You can overwrite this current branch but you would have to manually remove all the additional reviewers.

Oh, another note for rebasing--you can use our rebase bot to rebase onto viable/strict by commenting `@pytorchbot rebase -s`. https://github.com/pytorch/pytorch/wiki/Bot-commands

redmvc committed:
Add noncontiguous tensor logic to CUDA version

redmvc committed:
Fix order of a comment

redmvc committed:
Merge commit '50ce305' into less-strict-size-check-in-scatter

redmvc committed:
Change variable-size array implementation to be C++ compatible

redmvc committed:
Fix dtype declaration of index in documentation

redmvc committed:
Merge branch 'master' into less-strict-size-check-in-scatter

Zettelkasten commented:
Hey, is there still development on this? I would love to see this merged.

redmagnos commented:
@Zettelkasten It's still (as I understand it) undecided whether the desired behaviour is the logic I implemented (looping behaviour) or a simpler logic (overlapping behaviour) and until then this is on pause.
